### PR TITLE
Convert HELPER_METHOD_FRAME to QCalls (5/N)

### PR DESCRIPTION
**Pull request by jkotas**

### CODE REVIEW

File: src/coreclr/vm/comdelegate.cpp [hunk 1]
```diff
@@ -1000,36 +991,37 @@ FCIMPL5(FC_BOOL_RET, COMDelegate::BindToMethodName,
...
-    HELPER_METHOD_FRAME_END();
 
-    FC_RETURN_BOOL(pMatchingMethod != NULL);
-}
-FCIMPLEND
+    GCPROTECT_END();
+
+    END_QCALL;
 
+    return (pMatchingMethod != NULL);
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/comdelegate.cpp:999]: This is technically a C++ `bool` and not a `BOOL`. We should be consistent here with types.

File: src/coreclr/vm/comdelegate.cpp [hunk 2]
```diff
@@ -2110,46 +2095,40 @@ DELEGATEREF COMDelegate::CreateWrapperDelegate(DELEGATEREF delegate, MethodDesc*
...
-    HELPER_METHOD_FRAME_BEGIN_RET_2(refLeft, refRight);
+    BEGIN_QCALL;
 
-    MethodDesc* pMDLeft = GetMethodDesc(refLeft);
-    MethodDesc* pMDRight = GetMethodDesc(refRight);
+    GCX_COOP();
+
+    MethodDesc* pMDLeft = COMDelegate::GetMethodDesc(left.Get());
+    MethodDesc* pMDRight = COMDelegate::GetMethodDesc(right.Get());
     fRet = pMDLeft == pMDRight;
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/comdelegate.cpp:2126]:
```suggestion
    fRet = (pMDLeft == pMDRight) ? TRUE : FALSE;
```

File: src/coreclr/vm/commodule.cpp [hunk 3]
```diff
@@ -761,54 +728,50 @@ Object* GetTypesInner(Module* pModule)
                                              ClassLoader::ThrowIfNotFound,
                                              ClassLoader::PermitUninstDefOrRef);
         }
-        EX_CATCH_THROWABLE(&throwable);
+        EX_CATCH_THROWABLE(&gc.throwable);
 
-        if (throwable != 0) {
+        if (gc.throwable != 0) {
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/commodule.cpp:733]:
```suggestion
        if (gc.throwable != NULL) {
```

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeModule.cs [hunk 4]
```diff
@@ -15,21 +15,21 @@ internal sealed partial class RuntimeModule : Module
         internal RuntimeModule() { throw new NotSupportedException(); }
 
         #region FCalls
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeModule.cs:17]: Update or remove?

File: src/coreclr/vm/comdelegate.cpp (hunk 1 above)
Comment (jkotas) [reply src/coreclr/vm/comdelegate.cpp:999]: There is implicit conversion from `bool` to `BOOL`. We depend on this implicit conversion in thousands of places. Is depending on this implicit conversion bad for some reasons?

(I am fine with adding the explicit conversion in the two places you have commented on, but I do not understand why it helps.)

File: src/coreclr/vm/comutilnative.h [hunk 5]
```diff
@@ -207,6 +209,10 @@ extern "C" void* QCALLTYPE GCInterface_RegisterFrozenSegment(void *pSection, SIZ
 extern "C" void QCALLTYPE GCInterface_UnregisterFrozenSegment(void *segmentHandle);
 #endif // FEATURE_BASICFREEZE
 
+extern "C" int QCALLTYPE GCInterface_WaitForFullGCApproach(int millisecondsTimeout);
+
+extern "C" int QCALLTYPE GCInterface_WaitForFullGCComplete(int millisecondsTimeout);
```
Comment (elinor-fung) [inline src/coreclr/vm/comutilnative.h:214]: Remove the FCall declarations above?

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeModule.cs (hunk 4 above)
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeModule.cs:17]:
```suggestion
```

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeModule.cs [hunk 6]
```diff
@@ -15,21 +15,21 @@ internal sealed partial class RuntimeModule : Module
...
 
         internal RuntimeType[] GetDefinedTypes()
         {
-            return GetTypes(this);
+            RuntimeType[]? types = null;
+            RuntimeModule thisAsLocal = this;
+            GetTypes(new QCallModule(ref thisAsLocal), ObjectHandleOnStack.Create(ref types));
+            return types!;
         }
         #endregion
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeModule.cs:34]:
```suggestion
```

**Review by elinor-fung — APPROVED** [review]

File: src/coreclr/vm/comutilnative.h (hunk 5 above)
Comment (jkotas) [reply src/coreclr/vm/comutilnative.h:212]: Fixed

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeModule.cs (hunk 4 above)
Comment (jkotas) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeModule.cs]: Moved FCalls next to their uses. Also, deleted some unused fields.

File: src/coreclr/vm/comdelegate.cpp (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/vm/comdelegate.cpp:999]: > Is depending on this implicit conversion bad for some reasons?

No. It has no functional impact and as stated it is correct. My reasoning for it is the historical issue at the interop boundary where `bool` and `BOOL` are used interchangably and assumptions creep in. When I see P/Invoked methods I tend to make it explicit so it is clear in code what people are using. Most of the newer interop code uses `bool` throughout, but at the boundaries occasionally shifts to `BOOL` in an explicit way as a way to acknowledge the type being used.

In this case, I _think_ the caller could change the marshal to `UnmanagedType.U1` and completely avoid the normalization that typically occurs. The `DisableRuntimeMarshalling` does impact that but as a demonstration of "best practice" we've been explicit about types.

Comment (AaronRobinsonMSFT) [reply src/coreclr/vm/comdelegate.cpp:999]: I'm not pressing on it, but it is inconsistent with the other boolean returns in this PR. In fact, it is the only one that uses the implicit conversion.

Comment (jkotas) [reply src/coreclr/vm/comdelegate.cpp:999]: Thank you for the discussion. I think I am going to leave it as is. I do not see the explicit `? TRUE : FALSE` as adding clarity. I think it is fine to depend on implicit conversions.

> I think the caller could change the marshal to UnmanagedType.U1 and completely avoid the normalization

I see this just as a style question. There is no material perf difference between `BOOL` and `bool` for custom P/Invokes. C/C++ compilers are very good at optimizing the bool normalizations. Depending on the compiler and the architecture, the difference can be one extra instruction in the worst case. And the second normalization happens as part of LibraryImport marshalling for both `UnmanagedType.U1` and `UnmanagedType.Bool`. 

Outside tests, I only see 6 `UnmanagedType.U1` for custom P/Invokes in `System.Security.Cryptography.Native.Android`. All other custom P/Invokes in this repo use `UnmanagedType.Bool` in this repo.

**Pull request merged**